### PR TITLE
Mvapich2: Address issue with external MPI under Cray

### PR DESCRIPTION
**Pull request by jjellio**
The `mvapich2` package has some assumptions about how to setup external packages that can be problematic.

The issue is `if external_modules and 'cray' in external_modules[0]:`, which is the logic used to determine whether MPI's compilers will be set as `mpicc` or spack's wrapper.  With Cray, you should use spack's wrappers which point to cray's wrappers (not to mvapich's mpi compilers).

The problem with the logic used is it assumes the first module will have 'cray' in the name - which is necessarily true (and I am uses spack recipes provided by Cray that are breaking this assumption). 

I would like the maintainers to consider a more intuitive approach.

Given the package name is `mvapich2`, I propose search the modules for a module name containing `mvapich` or `cray-mvapich`. If this is the case, then almost certainly you are using the Cray toolchain and envs like `MPICC` and the spec's compiler `mpicc` should point to `spack_cc`

I've used this logic in the two areas where the issues arrises (setting up dependent packages and setting the compiler env)

This patch resolved some odd ball behavior I was observing on a Cray platform - that at first glance would be blamed on a bad package (pnetcdf in this case) - or a bug spack's external modules (there sorta is one) - which will be submitted in a separate issue.

Either way, I propose not using an undocumented assumption on how the user will list modules in their external declaration - if this patch can pass the regression tests, I'd like it considered for mainstream use!

@nithintsk @harisubramoni

### CODE REVIEW

File: var/spack/repos/builtin/packages/mvapich2/package.py [hunk 1]
```diff
@@ -347,11 +348,40 @@ def setup_dependent_build_environment(self, env, dependent_spec):
...
+        have_cray = False
+        if external_modules:
+            tty.msg('Detected external modules, will test for a'
+                    ' module that starts with {0}'.format(expected_prefix))
+            for module_name in external_modules:
+                if module_name.startswith(expected_prefix):
+                    have_cray = True
+                    msg = 'Found module with name {0} - assuming Cray environment'
+                    tty.msg(msg.format(module_name))
+        return have_cray
```
Comment (becker33) [inline var/spack/repos/builtin/packages/mvapich2/package.py]:
```suggestion
        expected_prefix = 'cray-mvapich'
        return any(m.startswith(expected_prefix) 
                          for m in external_modules)
```

File: var/spack/repos/builtin/packages/mvapich2/package.py [hunk 2]
```diff
@@ -347,11 +348,40 @@ def setup_dependent_build_environment(self, env, dependent_spec):
...
+
+        # use the this (since it's used elsewhere) to tag a package's message
+        pre = '{s.name}@{s.version} :'.format(s=self.spec)
+        if have_cray:
+            tty.msg(pre + 'Assuming Cray enviroment and setting MPI ENV'
+                    ' variables to spack wrappers')
+            tty.verbose(pre + 'Setting MPICC={0}'.format(spack_cc))
+            tty.verbose(pre + 'Setting MPICXX={0}'.format(spack_cxx))
+            tty.verbose(pre + 'Setting MPIF77={0}'.format(spack_fc))
+            tty.verbose(pre + 'Setting MPIF90={0}'.format(spack_fc))
```
Comment (becker33) [inline var/spack/repos/builtin/packages/mvapich2/package.py]: This should be less verbose once you're done with testing it.

**Review by becker33 — CHANGES_REQUESTED** [review]
This looks like a good idea. Cleaned it up a bit with some idiomatic python, and it needs to be less verbose for the user.

File: var/spack/repos/builtin/packages/mvapich2/package.py (hunk 2 above)
Comment (jjellio) [reply var/spack/repos/builtin/packages/mvapich2/package.py]: Is it possible to have them become `debug`, e.g., `spack -d` (is there a `tty.debug`)- I think `--verbose` should show it in the current incarnation.

I'm pushing back a little, because these messages were added because this is a deviation from the norm. It seems a bit redundant to list all wrappers, but if it only showed with `-d` or `-dd` that would enable really good info for bug reports.

To sort out what was going on, I ended up putting `traceback` call in the spec.py files - then I knew where these variables got set - my intuition blamed Cray or the modules - but it turned out to be a packaging thing.

File: var/spack/repos/builtin/packages/mvapich2/package.py (hunk 1 above)
Comment (jjellio) [reply var/spack/repos/builtin/packages/mvapich2/package.py]: Will this work? is `external_modules` promised to be defined? If it's `None`, then the for loop will throw (I think)

File: var/spack/repos/builtin/packages/mvapich2/package.py (hunk 2 above)
Comment (jjellio) [reply var/spack/repos/builtin/packages/mvapich2/package.py]: I've changed all output to be `tty.debug` - this keeps the console clean by default.

File: var/spack/repos/builtin/packages/mvapich2/package.py (hunk 1 above)
Comment (jjellio) [reply var/spack/repos/builtin/packages/mvapich2/package.py]: tentatively, I've keep my code organization. The primary value of the more verbose code is that info can be fed back to user (changed `tty.msg => tty.debug`.

IMO, that type of feedback is really important - particularly given the issues that seem to come up with modules.

**Comment by jjellio** [comment]
@becker33 : 
* I've addressed the verbosity issues. 
* As for code organization. I prefer having the ability to have debug output of reducing some of it to one-liner.
* Code in the setup portions got a bit more verbose because you can't read from `env` (you can from `os.environ`, but `env` is the spack `EnvironmentModification` type which does not provide subscripts `[]` or `get`).  Again, the code organization focuses on providing verbose messaging (if you add `-d`) - which I think is valuable given this involves cray/modules.

File: var/spack/repos/builtin/packages/mvapich2/package.py (hunk 2 above)
Comment (jjellio) [reply var/spack/repos/builtin/packages/mvapich2/package.py]: I've changed to `tty.debug` which keeps the console clean by default.

For the setup routine, you can't read `env` so I had to do a little extra leg work. The code organization allows for printing what spack does.

File: var/spack/repos/builtin/packages/mvapich2/package.py (hunk 1 above)
Comment (jjellio) [reply var/spack/repos/builtin/packages/mvapich2/package.py]: I've kept my code for now - it allows messages to get printed.

Moreover, You'd need to have a test for `external_modules` anyway - if it's none you choke. I suppose I may need a string test for `external_modules` or an assertion that `external_modules` is a list.

I'll add a `six` strings type test and repush.

Resolving this for now.

**Comment by skosukhin** [comment]
@jjellio Although I agree that the current implementation is a bit awkward, I suggest that you take a look at #20076 (and, of course, #20663) and do the same thing for the `mvapich2` package. That should eliminate the ambiguity right away.

**Comment by jjellio** [comment]
@becker33 Can you re-review - I think I got everything you wanted checked out

**Comment by alalazo** [comment]
This is an issue with Cray and modules. Should we close it as deprecated since we now have the manifest file on Cray systems?

**Comment by becker33** [comment]
Since the separate `cray` platform is now deprecated in favor of using the new cray mpi wrappers, I will close this PR.
